src/main/java/org/folio/auth/authtokenmodule/apis/RouteApi.java

```diff
@@ -179,43 +201,95 @@ private void handleRefresh(RoutingContext ctx) {
     try {
       String content = ctx.getBodyAsString();
       JsonObject requestJson = new JsonObject(content);
-      String encryptedJWE = requestJson.getString("refreshToken");
-      var context = new TokenValidationContext(ctx.request(), tokenCreator, encryptedJWE);
+      String encryptedJWE = requestJson.getString(REFRESH_TOKEN);
+
+      String tenant = ctx.request().headers().get(XOkapiHeaders.TENANT);
+      var refreshTokenStore = new RefreshTokenStore(vertx, tenant);
+
+      var context =
+         new TokenValidationContext(ctx.request(), tokenCreator, encryptedJWE, refreshTokenStore);
       Future<Token> tokenValidationResult = Token.validate(context);
 
-      tokenValidationResult.onFailure(h -> {
-        handleTokenValidationFailure(h, ctx);
-      });
+      tokenValidationResult.onFailure(e -> handleTokenValidationFailure(e, ctx));
 
       tokenValidationResult.onSuccess(token -> {
-        String username = token.getClaims().getString("sub");
-        String userId = token.getClaims().getString("user_id");
-        String tenant = token.getClaims().getString("tenant");
-
         try {
-          // TODO To do RTR we need to return both a new AT and a new RT here.
+          String username = token.getClaims().getString("sub");
+          String userId = token.getClaims().getString(USER_ID);
+
           String at = new AccessToken(tenant, username, userId).encodeAsJWT(tokenCreator);
-          JsonObject responseObject = new JsonObject().put("token", at);
-          endJson(ctx, 201, responseObject.encode());
+          JsonObject responseObject = new JsonObject().put(ACCESS_TOKEN, at);
+
+          String address = ctx.request().remoteAddress().host();
+          var rt = new RefreshToken(tenant, username, userId, address);
+          responseObject.put(REFRESH_TOKEN, rt.encodeAsJWE(tokenCreator));
+
+        // Save the RT to track one-time use.
+        new RefreshTokenStore(vertx, tenant).saveToken(rt).onComplete(x ->
```

Comment:
Is it possible to use the usual pattern with .compose and error handling at the end avoiding the try-catch?

```
Token.validate(context)
.compose(token -> {
  ...
  return new RefreshTokenStore(vertx, tenant).saveToken(rt)
      .map(responseObject);
})
.onSuccess(responseObject -> endJson(ctx, 201, responseObject.encode()));
.onFailure(e -> ...);
```

Reply:
Done. See new `returnToken` method.